### PR TITLE
One step migration

### PR DESCRIPTION
**Pull request by yves-da**
Documentation for https://github.com/DACH-NY/canton/issues/18541

### CODE REVIEW

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 1]
```diff
@@ -645,6 +651,58 @@ the :ref:`transfer command <transfer.execute>`, contracts can be moved over to t
...
+
+.. _one_step_migration:
+
+One Step Migration
+------------------
+A *one step migration* intends to :ref:`upgrade the binary <upgrade_canton_binary>` and
+:ref:`change the protocol version <change_pv>` in one step. It is a combination of the
+steps documented in :ref:`Hard Synchronization Domain Connection Upgrade <hard_sync_domain_upgrade>`
+and :ref:`Upgrade Canton Binary <upgrade_canton_binary>` from a operational point of
+view. Thus, we assume strong familiarity with aforementioned sections.
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:666]:
```suggestion
view. Thus, strong understanding of the aforementioned sections is required.
```
Hi, according to our style guide, we're steering away from the "we" or "one" voice and using the "you" voice instead, when necessary. Thank you.

**Review by cathyjung-da — CHANGES_REQUESTED** [review]

**Review by cathyjung-da — APPROVED** [review]

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 2]
```diff
@@ -37,22 +37,30 @@ individual notes for each version.
...
 - A set of configuration files describing the node that is being run.
 - An optional bootstrap script passed via ``--boostrap``, which runs on startup.
 - A database (with a specific schema), holding the data of the node.
 
-Therefore, to upgrade the node, you will need to not only replace the jar, but also test that
-the configuration files can still be parsed by the new process, that the bootstrap script you
-are using is still working, and you need to upgrade the database schema.
+Therefore, to upgrade the node, you need to
+
+#. replace the Canton binary (which contains the Canton JAR)
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:60]:
```suggestion
#. Replace the Canton binary (which contains the Canton JAR).
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 3]
```diff
@@ -37,22 +37,30 @@ individual notes for each version.
...
+#. test that the configuration files can still be parsed by the new process
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:61]:
```suggestion
#. Test that the configuration files can still be parsed by the new process.
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 4]
```diff
@@ -37,22 +37,30 @@ individual notes for each version.
...
+#. test that the bootstrap script you are using is still working
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:62]:
```suggestion
#. Test that the bootstrap script you are using is still working.
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 5]
```diff
@@ -37,22 +37,30 @@ individual notes for each version.
...
+#. upgrade the database schema
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:63]:
```suggestion
#. Upgrade the database schema.
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 6]
```diff
@@ -67,7 +75,7 @@ Preparation
 ~~~~~~~~~~~
 
 First, please download the new Canton binary that you want to upgrade to and store it on the test
-system where you want to test the upgrade process first.
+system where you want to test the upgrade process.
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:78]:
```suggestion
system where you plan to test the upgrade process.
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 7]
```diff
@@ -492,12 +514,17 @@ Therefore, the protocol upgrade process boils down to:
 
 Note: to use the same database with different schemas for the old and the new sync domain, set the `currentSchema` either in the JDBC URL or as a parameter in `storage.config.properties`.
 
+.. _hard_sync_domain_upgrade:
+
 Hard Synchronization Domain Connection Upgrade
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+.. warning::
+  Ensure that you have appropriate backups in place and that you have tested this procedure before applying
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:523]:
```suggestion
  Ensure that you have appropriate backups in place and have tested this procedure before applying
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 8]
```diff
@@ -492,12 +514,17 @@ Therefore, the protocol upgrade process boils down to:
...
+  it to your production system.
+
 A hard sync domain connection upgrade can be performed using the :ref:`respective migration command <repair.migrate_domain>`.
-Again, please ensure that you have appropriate backups in place and that you have tested this procedure before applying
-it to your production system. You will have to enable these commands using a special config switch:
+You will have to enable this command using a special config switch:
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:527]:
```suggestion
You must enable this command using a special config switch:
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 9]
```diff
@@ -555,7 +582,7 @@ Now, using this configuration object, we can trigger the hard sync domain connec
     .. success:: participant.repair.migrate_domain("olddomain", config)
 
 This command will register the new sync domain and re-associate the contracts tied to ``olddomain`` to
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:584]:
```suggestion
This command registers the new sync domain and re-associate the contracts tied to ``olddomain`` to
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 10]
```diff
@@ -645,6 +672,58 @@ the :ref:`transfer command <transfer.execute>`, contracts can be moved over to t
...
+------------------
+A *one step migration* intends to :ref:`upgrade the binary <upgrade_canton_binary>` and
+:ref:`change the protocol version <change_pv>` in one step. It is a combination of the
+steps documented in :ref:`Hard Synchronization Domain Connection Upgrade <hard_sync_domain_upgrade>`
+and :ref:`Upgrade Canton Binary <upgrade_canton_binary>` from a operational point of
+view. Thus, strong understanding of the aforementioned sections is required.
+
+.. note::
+
+    There is no need for a one step migration when upgrading from one release version to another
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:691]:
```suggestion
    There is no need for a one-step migration when upgrading from one release version to another,
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 11]
```diff
@@ -645,6 +672,58 @@ the :ref:`transfer command <transfer.execute>`, contracts can be moved over to t
...
+    while both release versions support the protocol version that your current sync domain is using.
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:692]:
```suggestion
    as long as both release versions support the protocol version that your current sync domain uses.
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 12]
```diff
@@ -645,6 +672,58 @@ the :ref:`transfer command <transfer.execute>`, contracts can be moved over to t
...
 the old sync domain to be decommissioned and turned off.
 
-The soft upgrade path provides a smooth user experience that does not require a hard migration of the sync domain connection to be 
+The soft upgrade path provides a smooth user experience that does not require a hard migration of the sync domain connection to be
 coordinated across all participants. Instead, participants upgrade individually, whenever they
 are ready, allowing them to reverse the process if needed.
+
+.. _one_step_migration:
+
+One Step Migration
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:681]:
```suggestion
One-Step Migration
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 13]
```diff
@@ -645,6 +672,58 @@ the :ref:`transfer command <transfer.execute>`, contracts can be moved over to t
...
+------------------
+A *one step migration* intends to :ref:`upgrade the binary <upgrade_canton_binary>` and
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:683]:
```suggestion
A *one-step migration* intends to :ref:`upgrade the binary <upgrade_canton_binary>` and
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 14]
```diff
@@ -645,6 +672,58 @@ the :ref:`transfer command <transfer.execute>`, contracts can be moved over to t
...
+and :ref:`Upgrade Canton Binary <upgrade_canton_binary>` from a operational point of
+view. Thus, strong understanding of the aforementioned sections is required.
+
+.. note::
+
+    There is no need for a one step migration when upgrading from one release version to another
+    while both release versions support the protocol version that your current sync domain is using.
+    Follow the steps for :ref:`upgrading the Canton binary <upgrade_canton_binary>` instead.
+
+The one step migration covers an upgrade from Canton version 2.3 and following minor versions
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:695]:
```suggestion
The one-step migration covers an upgrade from Canton version 2.3 and following minor versions
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 15]
```diff
@@ -645,6 +672,58 @@ the :ref:`transfer command <transfer.execute>`, contracts can be moved over to t
...
+up to this minor release version. Additionally, it upgrades the protocol version supported by
+these prior releases to the latest protocol version supported by this minor release version.
+
+Every upgrade requires specific and thorough testing before applying it to a particular production
+environment. Even though the one step migration process has been tested through automatic and manual
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:700]:
```suggestion
environment. Even though the one-step migration process has been tested through automatic and manual
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 16]
```diff
@@ -645,6 +672,58 @@ the :ref:`transfer command <transfer.execute>`, contracts can be moved over to t
...
+tests and its steps are known to work, additional measures and configuration may be required to
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:701]:
```suggestion
tests, and its steps are known to work, additional measures and configuration may be required to
```

**Review by cathyjung-da — CHANGES_REQUESTED** [review]

File: docs/2.9.0/docs/canton/includes/snippet_data/usermanual/connectivity.json
Comment (yves-da) [inline docs/2.9.0/docs/canton/includes/snippet_data/usermanual/connectivity.json:1]: Note: Generated, ignore.

File: docs/2.9.0/docs/canton/includes/snippet_data/usermanual/identity_management.json
Comment (yves-da) [inline docs/2.9.0/docs/canton/includes/snippet_data/usermanual/identity_management.json:1]: Note: Generated, ignore.

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst
Comment (yves-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:1]: Note: Actual changes, please review.

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 17]
```diff
@@ -194,6 +205,27 @@ The ping command creates two contracts between the admin parties, then exercises
...
 ~~~~~~~~~~~~~~~~~~~~~~
 
+.. _upgrade_to_2.9:
+
+Upgrade to Release 2.9
+^^^^^^^^^^^^^^^^^^^^^^
+
+TODO
+
+Other
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:215]: Find more precise title ? It appears as "Other" in the ToC which is a bit weird

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 18]
```diff
@@ -194,6 +205,27 @@ The ping command creates two contracts between the admin parties, then exercises
...
+"""""
+Version 2.9 adds a new repair command :ref:`participant.repair.purge_deactivated_domain <repair.purge_deactivated_domain>`
+to delete data of a defunct domain.
+
+We recommend to use this command to remove remaining but useless data of previous
+sync domains that have been migrated using the :ref:`participant.repair.migrate_domain <repair.migrate_domain>`
+command.
+
+Note that the ``migrate_domain`` command in 2.9 now automatically removes such data,
+but only for the sync domain on which it has been invoked on.
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:225]: The details of the changes are in the release notes. In that page, we only list changes that require changes from the end users (adapt config, change a console command, ...) Since the purge_deactivated_domain is not needed for the migration (invoked transparently), I'd remove this new section for now.

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 19]
```diff
@@ -476,35 +519,59 @@ side by side with the old sync domain process.
...
 
-Please note that currently, the sync domain ID cannot be preserved during upgrades.
-The new sync domain must have a different sync domain ID because the participant
-associates a sync domain connection with a sync domain ID, and that association
-must be unique.
+Wit that, the protocol upgrade process boils down to:
+
+#. **Deploy a new sync domain with the new protocol version**
+
+   Deploy a new sync domain next to the old sync domain and ensure that the new sync
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:526]:
```suggestion
   Deploy a new sync domain and ensure that the new sync
```
not sure what "next to" means in context of domains.

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 20]
```diff
@@ -476,35 +519,59 @@ side by side with the old sync domain process.
...
+   domain is using the desired protocol version.
+
+   Also make sure to use different databases (or at least different schemas in the same database)
+   for the sync domain services (mediator, sequencer node, and topology manager), channel names,
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:530]:
```suggestion
   for the sync domain services (domain node, mediator, sequencer node, and topology manager), channel names,
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 21]
```diff
@@ -476,35 +519,59 @@ side by side with the old sync domain process.
...
+
+   Instruct the participants individually using the :ref:`hard sync domain migration<hard_sync_domain_upgrade>`
+   to use the new sync domain.
+
+.. note::
+
+    Currently, the sync domain ID cannot be preserved during upgrades.
+    The new sync domain must have a different sync domain ID because the participant
+    associates a sync domain connection with a sync domain ID, and that association
+    must be unique.
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:546]: As usual, I would make how-to guides as short as possible. Consider removing the second sentence which answers to a WHY question.

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 22]
```diff
@@ -514,28 +581,34 @@ You can do that by either controlling your applications, or by
     .. assert:: { participant.domains.connect_local(olddomain); true }
     .. success:: participant.resources.set_resource_limits(ResourceLimits(Some(0), Some(0)))
 
-This will reject all commands and finish processing the pending commands. Once you are sure that
-your participant node is idle, disconnect the participant node from the old sync domain
-connection:
+This rejects all commands and finishes processing the pending commands. Once you are sure that
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:584]:
```suggestion
This rejects any new command and finishes processing the pending commands. Once you are sure that
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 23]
```diff
@@ -514,28 +581,34 @@ You can do that by either controlling your applications, or by
...
 
-Test that the sync domain is disconnected by checking the list of active connections:
+Test that the participant is disconnected from the sync domain by checking the list of active connections:
 
 .. snippet:: migrating_protocol
     .. success:: participant.domains.list_connected()
     .. assert:: participant.domains.list_connected().isEmpty
 
-This is a good time to perform a backup of the database before proceeding:
+This is a good time to perform a backup of the database before proceeding.
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:596]: Why this change ?

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 24]
```diff
@@ -514,28 +581,34 @@ You can do that by either controlling your applications, or by
...
 
 .. code:: sql
 
     CREATE DATABASE newdb WITH TEMPLATE originaldb OWNER dbuser;
 
+.. warning::
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:602]: The fact that the warning shows before the invocation of the command but refers to it makes the reading a bit difficult. I'd suggest to rephrase slightly:
```
.. warning::

    Note that the hard domain migration modifies the participant's data storage.
    Without a database backup for your participant a potential recovery becomes significantly
    more difficult.
```
or even
```
.. warning::

    Note that the following steps modify participant's data storage.
    Without a database backup for your participant a potential recovery becomes significantly
    more difficult.
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 25]
```diff
@@ -554,10 +627,16 @@ Now, using this configuration object, we can trigger the hard sync domain connec
 .. snippet:: migrating_protocol
     .. success:: participant.repair.migrate_domain("olddomain", config)
 
-This command will register the new sync domain and re-associate the contracts tied to ``olddomain`` to
-the new sync domain.
+This command registers the new sync domain and re-associate the contracts tied to ``olddomain`` to
+the new sync domain. In addition some data specific to the old domain are automatically deleted.
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:631]:
```suggestion
the new sync domain. In addition some data specific to the old domain is automatically deleted.
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 26]
```diff
@@ -554,10 +627,16 @@ Now, using this configuration object, we can trigger the hard sync domain connec
...
+
+.. warning::
+
+    The command ``repair.migrate_domain`` features a ``force`` flag. If you intentionally set it to
+    ``true`` then you force a sync domain migration ignoring any pending transactions. Forcing a
+    migration may result in a ledger fork. Use this flag as a last resort only.
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:637]: What is the reason/purpose to advertise this here ?

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 27]
```diff
@@ -605,46 +683,78 @@ the upgrade test to understand the necessary downtime required for the migration
...
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+.. _one_step_migration:
 
-.. note ::
+One-Step Migration
+------------------
+A *one-step migration* intends to :ref:`upgrade the binary <upgrade_canton_binary>` and
+:ref:`change the protocol version <change_pv>` in one step. It is a combination of the
+steps documented in :ref:`upgrade_canton_binary` and the :ref:`hard_sync_domain_upgrade`.
+Thus, strong understanding of these sections is required.
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:693]: I'd remove that section: for me the user should be able to follow the steps **carefully** without necessarily understanding everything which is behind.

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 28]
```diff
@@ -605,46 +683,78 @@ the upgrade test to understand the necessary downtime required for the migration
...
-soft sync domain connection upgrade is operationally much simpler, and can be leveraged using multi-sync-domain
-support (which exists as a pre-alpha feature only for now). By turning off non-unique contract keys,
-participants can connect to multiple sync domains and transfer contracts between sync domains. This allows us to avoid using the ``repair.migrate_domain`` step.
+.. note::
 
-Assuming the same setup as before, where the participant is connected to the old sync domain,
-we can just connect it to the new sync domain
+    There is no need for a one-step migration when upgrading from one release version to another,
+    as long as both release versions support the protocol version that your current sync domain uses.
+    Follow the steps for :ref:`upgrading the Canton binary <upgrade_canton_binary>` instead.
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:703]: Add reference to the table which is in https://docs.daml.com/canton/usermanual/versioning.html#canton-protocol-version ?

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 29]
```diff
@@ -605,46 +683,78 @@ the upgrade test to understand the necessary downtime required for the migration
...
-    .. assert:: { participant.start(); true }
-    .. assert:: { participant.domains.connect_local(olddomain); true }
-    .. success:: participant.domains.connect_local(newdomain)
-    .. assert:: participant.domains.list_connected().map(_.domainAlias.unwrap).toSet == Set("newdomain", "olddomain")
+.. warning::
+
+    Every upgrade requires specific and thorough testing before applying it to a particular production
+    environment. Even though the one-step migration process has been tested through automatic and manual
+    tests, and its steps are known to work, additional measures and configuration may be required to
+    address the peculiarities of your environment.
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:710]: :+1:

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 30]
```diff
@@ -605,46 +683,78 @@ the upgrade test to understand the necessary downtime required for the migration
...
+:ref:`2.4 <upgrade_to_2.4>`, :ref:`2.5 <upgrade_to_2.5>`, :ref:`2.6 <upgrade_to_2.6>`,
+:ref:`2.7 <upgrade_to_2.7>` or :ref:`2.8 <upgrade_to_2.8>` release and protocol
+version 3 or 4 to a new sync domain running on the 2.9 release and protocol version 5:
+
+.. note::
+
+    We recommend using protocol version 5 even though 2.9 supports other protocol versions.
+
+#. Start a new sync domain running protocol version 5
+#. :ref:`Halt activity on the current sync domain <halt-activity-current-domain>`
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:724]: Do we want to use old/new terminology ? (so current->old)

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 31]
```diff
@@ -605,46 +683,78 @@ the upgrade test to understand the necessary downtime required for the migration
...
+#. Wait for pending transactions to complete or time out
+#. Backup the current sync domain including participants
+#. Participants: Upgrade the binary to 2.9
+#. Participants: :ref:`Enable repair commands <enable-repair-commands>`
+#. All nodes: :ref:`Test the configuration <test-your-config>` considering these :ref:`details <test-your-config-details>`
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:729]:
```suggestion
#. All nodes: :ref:`Test the configuration <test-your-config>` considering this :ref:`additional change <test-your-config-details>` if your old domain runs protocol version 3.
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 32]
```diff
@@ -605,46 +683,78 @@ the upgrade test to understand the necessary downtime required for the migration
...
+#. All nodes: :ref:`Apply the DB migrations <migrating_the_database>`
+#. Participants: Connect to the new sync domain, then disconnect from all domains
+#. Participants: Invoke :ref:`command<repair.migrate_domain>` ``repair.migrate_domain``
+#. Participants: Reconnect to the new sync domain
+#. Resume activity on the new sync domain
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:734]: Activity on the new domain is never halted, so you can drop this point.
What can be added is the the old domain can be decomissioned.

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 33]
```diff
@@ -605,46 +683,78 @@ the upgrade test to understand the necessary downtime required for the migration
...
+
+Halt activity on the current sync domain
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+For a sync domain running on protocol version 3, set the resource limits to 0 on each participant.
+
+.. code-block:: text
+
+    participant.resources.set_resource_limits(ResourceLimits(Some(0), Some(0)))
+
+Otherwise, set the :ref:`dynamic domain parameter <dynamic_domain_parameters>` ``maxRatePerParticipant`` to 0.
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:746]: I would put this one first: "For a sync domain running protocol version 4 and above, ..."

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 34]
```diff
@@ -605,46 +683,78 @@ the upgrade test to understand the necessary downtime required for the migration
...
 
-Give the new connection precedence over the old connection by changing the ``priority`` flag of the new
-sync domain connection:
+.. _test-your-config-details:
 
-.. snippet:: soft_migration_with_transfer
-    .. success:: participant.domains.modify("newdomain", _.copy(priority=10))
+Test the configuration - Additional details
```
Comment (rgugliel-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:750]:
```suggestion
Test the configuration - Additional change if you were running protocol version 3
```

**Review by rgugliel-da — APPROVED** [review]
LGTM
Thanks for the write up!

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 35]
```diff
@@ -476,35 +519,59 @@ side by side with the old sync domain process.
...
+.. note::
+
+    Currently, the sync domain ID cannot be preserved during upgrades.
+    The new sync domain must have a different sync domain ID because the participant
+    associates a sync domain connection with a sync domain ID, and that association
+    must be unique.
+
+.. note::
+
+    To use the same database with different schemas for the old and the new sync domain,
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:550]:
```suggestion
    To use the same database with different schemas for the old and the new sync domains,
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 36]
```diff
@@ -476,35 +519,59 @@ side by side with the old sync domain process.
...
+
+You must enable this command using a special config switch:
+
+.. code:: text
 
     canton.features.enable-repair-commands=yes
 
-The process of a hard migration is quite straightforward. Assuming that we have several participants,
-all connected to a sync domain named ``olddomain``, then ensure that there are no pending transactions.
+Assuming that we have several participants, all connected to a sync domain named ``olddomain``,
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:573]:
```suggestion
Assuming that you have several participants all connected to a sync domain named ``olddomain``,
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 37]
```diff
@@ -476,35 +519,59 @@ side by side with the old sync domain process.
...
+then ensure that there are no pending transactions.
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:574]:
```suggestion
ensure that there are no pending transactions.
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 38]
```diff
@@ -514,28 +581,34 @@ You can do that by either controlling your applications, or by
...
 
 .. code:: sql
 
     CREATE DATABASE newdb WITH TEMPLATE originaldb OWNER dbuser;
 
+.. warning::
+
+    Note that the hard domain migration :ref:`command<repair.migrate_domain>` ``repair.migrate_domain``
+    modifies the participant's data storage.
+    Without a database backup for your participant a potential recovery becomes significantly
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:606]:
```suggestion
    Without a database backup for your participant, a potential recovery becomes significantly
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst (hunk 25 above)
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:631]:
```suggestion
the new sync domain. In addition, some data specific to the old domain is automatically deleted.
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 39]
```diff
@@ -554,10 +627,16 @@ Now, using this configuration object, we can trigger the hard sync domain connec
...
 
-This command will register the new sync domain and re-associate the contracts tied to ``olddomain`` to
-the new sync domain.
+This command registers the new sync domain and re-associate the contracts tied to ``olddomain`` to
+the new sync domain. In addition some data specific to the old domain are automatically deleted.
+
+.. warning::
+
+    The command ``repair.migrate_domain`` features a ``force`` flag. If you intentionally set it to
+    ``true`` then you force a sync domain migration ignoring any pending transactions. Forcing a
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:636]:
```suggestion
    ``true``, it forces a sync domain migration, ignoring any pending transactions. Forcing a
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 40]
```diff
@@ -554,10 +627,16 @@ Now, using this configuration object, we can trigger the hard sync domain connec
...
     .. success:: participant.repair.migrate_domain("olddomain", config)
 
-This command will register the new sync domain and re-associate the contracts tied to ``olddomain`` to
-the new sync domain.
+This command registers the new sync domain and re-associate the contracts tied to ``olddomain`` to
+the new sync domain. In addition some data specific to the old domain are automatically deleted.
+
+.. warning::
+
+    The command ``repair.migrate_domain`` features a ``force`` flag. If you intentionally set it to
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:635]:
```suggestion
    The command ``repair.migrate_domain`` has a ``force`` flag. If it is explicitly set to
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 41]
```diff
@@ -605,46 +683,78 @@ the upgrade test to understand the necessary downtime required for the migration
...
+One-Step Migration Recipe for 2.9
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+General recipe to migrate from a sync domain running on a :ref:`2.3 <upgrade_to_2.3>`,
+:ref:`2.4 <upgrade_to_2.4>`, :ref:`2.5 <upgrade_to_2.5>`, :ref:`2.6 <upgrade_to_2.6>`,
+:ref:`2.7 <upgrade_to_2.7>` or :ref:`2.8 <upgrade_to_2.8>` release and protocol
+version 3 or 4 to a new sync domain running on the 2.9 release and protocol version 5:
+
+.. note::
+
+    We recommend using protocol version 5 even though 2.9 supports other protocol versions.
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:721]:
```suggestion
    Although version 2.9 supports other protocol versions, it is recommended to use protocol version 5.
```

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst [hunk 42]
```diff
@@ -605,46 +683,78 @@ the upgrade test to understand the necessary downtime required for the migration
...
-sync domain connection:
+.. _test-your-config-details:
 
-.. snippet:: soft_migration_with_transfer
-    .. success:: participant.domains.modify("newdomain", _.copy(priority=10))
+Test the configuration - Additional details
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-You can check the priority settings of the sync domains using
+When migrating from protocol version 3 which uses an unauthenticated contract id scheme, you may need to
```
Comment (cathyjung-da) [inline docs/2.9.0/docs/canton/usermanual/upgrading.rst:753]:
```suggestion
When migrating from protocol version 3, which uses an unauthenticated contract id scheme, you may need to
```

**Review by cathyjung-da — CHANGES_REQUESTED** [review]

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst (hunk 18 above)
Comment (yves-da) [reply docs/2.9.0/docs/canton/usermanual/upgrading.rst]: > Since the purge_deactivated_domain is not needed for the migration (invoked transparently), I'd remove this new section for now.

I added this section specifically to point out that customers should run the command for domains which they have migrated before as I think their data still lingers in the data store. When doing a hard domain migration only the old (source) domain gets cleaned up, but other inactive (deactivated) domains remain as is.

We added this command specifically for that purpose.

File: docs/2.9.0/docs/canton/usermanual/upgrading.rst (hunk 26 above)
Comment (yves-da) [reply docs/2.9.0/docs/canton/usermanual/upgrading.rst]: Over-documentation -> removed.

**Review by cathyjung-da — APPROVED** [review]
